File: python/백준/자료구조/덱, 큐/3190 뱀.py
import sys
from collections import deque

input = sys.stdin.readline

N = int(input())
K = int(input())

apples = []

for _ in range(K):
    x, y = map(int, input().split())
    apples.append((x, y))
    
L = int(input())

commands = []

for _ in range(L):
    second, direction = map(str, input().split())
    commands.append((int(second), direction))
    
answer = 0

directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]

def change_direction(current_direction, next_direction):
    if next_direction == 'D':
        return (current_direction + 1) % 4
    elif next_direction == 'L':
        return (current_direction - 1) % 4
        
snake = deque([(1, 1)])
current_direction = 0

while True:
    answer += 1
    
    current_head_x, current_head_y = snake[-1]
    
    next_head_x, next_head_y = current_head_x + directions[current_direction][0], current_head_y + directions[current_direction][1]
    
    if (next_head_x, next_head_y) in snake:
        break
    
    if 0 < next_head_x <= N and 0 < next_head_y <= N:
        snake.append((next_head_x, next_head_y))
    else:
        break
    
    if (next_head_x, next_head_y) in apples:
        apples.remove((next_head_x, next_head_y))
    else:
        snake.popleft()
        
    if commands:
        if commands[0][0] == answer:
            current_direction = change_direction(current_direction, commands[0][1])
            commands.pop(0)
            
print(answer)

File: cpp/백준/자료구조/덱, 큐/3190 뱀.cpp
#include <iostream>
#include <deque>
#include <vector>
#include <set>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int N;
    cin >> N;
    int K;
    cin >> K;

    set<pair<int, int>> apples;
    for (int i = 0; i < K; ++i) {
        int x, y;
        cin >> x >> y;
        apples.insert({x, y});
    }

    int L;
    cin >> L;

    deque<pair<int, char>> commands;
    for (int i = 0; i < L; ++i) {
        int second;
        char direction;
        cin >> second >> direction;
        commands.push_back({second, direction});
    }

    int answer = 0;
    int dx[] = {0, 1, 0, -1};
    int dy[] = {1, 0, -1, 0};

    deque<pair<int, int>> snake;
    snake.push_back({1, 1});
    set<pair<int, int>> snake_set;
    snake_set.insert({1, 1});
    int current_direction = 0;

    while (true) {
        answer += 1;

        auto [current_head_x, current_head_y] = snake.back();
        int next_head_x = current_head_x + dx[current_direction];
        int next_head_y = current_head_y + dy[current_direction];

        if (snake_set.count({next_head_x, next_head_y})) break;

        if (next_head_x > 0 && next_head_x <= N && next_head_y > 0 && next_head_y <= N) {
            snake.push_back({next_head_x, next_head_y});
            snake_set.insert({next_head_x, next_head_y});
        } else {
            break;
        }

        if (apples.count({next_head_x, next_head_y})) {
            apples.erase({next_head_x, next_head_y});
        } else {
            snake_set.erase(snake.front());
            snake.pop_front();
        }

        if (!commands.empty() && commands.front().first == answer) {
            char nd = commands.front().second;
            if (nd == 'D') current_direction = (current_direction + 1) % 4;
            else if (nd == 'L') current_direction = (current_direction - 1 + 4) % 4;
            commands.pop_front();
        }
    }

    cout << answer << "\n";
    return 0;
}
